woocommerce/google-listings-and-ads
# Milestone 1 Tree Select Component

### Changes proposed in this Pull Request:

Milestone 1 in TreeSelectComponent

Minimum Viable Product to demonstrate must-have functionalities, basic UI structure, and API. The component will initially be developed in Google Listings and Ads repo for easier consumption by the projects.

More info: pcTzPl-fg-p2

- Specify the data structure needed for the component. Each node should have a unique key or value.
- Display the data in tree structure in a dropdown select with checkboxes.
- Checkboxes should have two states for now: checked and unchecked
- The component must have `value` and `onChange` props. `value` should be an array of keys. `onChange` is a callback function passing the selected values (array of keys).
- Display selected values as tags in the component.
- Disabled state. Showing the selected tags but not allowing to change the selection. 
- Allow clicking on “X” in the tag to remove the value.
- Basic styling

### Screenshots:

https://user-images.githubusercontent.com/5908855/154451422-622f81d3-75e5-414c-bd01-93de1a6e1376.mov

### Detailed test instructions:

1. Checkout the branch
2. Go to `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fcomponent-test` 
3. You can see component TreeSlectControl with the label `Select Country` and the placeholder `Select`
4. Clicking on "Select" expands the TreeSlector and shows text cursor instead of the placeholder  
5. Tree Shows Europe (It, Es, Fr) and Asia (Ch, Jp)
6. Clicking on the country checks it and shows the tag
7. Clicking again unchecks it and removes the tag
8. Clicking on the "x" button in a tag, removes the tag
9. If you click on all the countries, it automatically selects the Continent and adds all of the countries on it
10. If you click the Continent it deselects/select all the countries
11. If some country is not checked, the continent is not checked
12. Clicking out the selector close the tree
13. Clicking again in the selector expands the tree and preserves the selections
14. Pay attention inside the code to the logic we use to define the data structure, render the tree and tags as well as removing, changing elements.
